crates/holochain/src/core/state/cascade.rs

```diff
@@ -303,31 +307,137 @@ where
+            highest_observed,
+            valid_activity,
+            ..
+        } = agent_activity;
+        match valid_activity {
+            Activity::Full(headers) => {
+                let hashes = headers
+                    .into_iter()
+                    .map(|shh| (shh.header().header_seq(), shh.header_address().clone()));
+                cache_data.meta.register_activity_sequence(&agent, hashes)?;
+            }
+            Activity::Hashes(hashes) => {
+                let hashes = hashes.into_iter();
+                cache_data.meta.register_activity_sequence(&agent, hashes)?;
+            }
+            Activity::NotRequested => (),
+        };
+        cache_data.meta.register_activity_status(&agent, status)?;
+        if let Some(highest_observed) = highest_observed {
+            cache_data
+                .meta
+                .register_activity_observed(&agent, highest_observed)?;
+        }
+        Ok(())
+    }
+
+    fn update_stores(&mut self, element: Element) -> CascadeResult<()> {
         let cache_data = ok_or_return!(self.cache_data.as_mut());
-        let op_lights = produce_op_lights_from_elements(vec![&element]).await?;
+        let op_lights = produce_op_lights_from_elements(vec![&element])?;
         let (shh, e) = element.into_inner();
-        cache_data.element.put(shh, option_entry_hashed(e).await)?;
+        cache_data.element.put(shh, option_entry_hashed(e))?;
         for op in op_lights {
             integrate_single_metadata(op, cache_data.element, cache_data.meta)?
         }
         Ok(())
     }
 
     #[instrument(skip(self, elements))]
-    async fn update_stores_with_element_group(
+    fn update_stores_with_element_group(
         &mut self,
         elements: ElementGroup<'_>,
     ) -> CascadeResult<()> {
         let cache_data = ok_or_return!(self.cache_data.as_mut());
-        let op_lights = produce_op_lights_from_element_group(&elements).await?;
+        let op_lights = produce_op_lights_from_element_group(&elements)?;
         cache_data.element.put_element_group(elements)?;
         for op in op_lights {
             integrate_single_metadata(op, cache_data.element, cache_data.meta)?
         }
         Ok(())
     }
 
+    fn put_element_in_cache(&mut self, response: GetElementResponse) -> CascadeResult<()> {
+        match response {
+            // Has header
+            GetElementResponse::GetHeader(Some(we)) => {
+                let (element, delete) = we.into_element_and_delete();
+                self.update_stores(element)?;
+
+                if let Some(delete) = delete {
+                    self.update_stores(delete)?;
+                }
+            }
+            // Doesn't have header but not because it was deleted
+            GetElementResponse::GetHeader(None) => (),
+            r => {
+                error!(
+                    msg = "Got an invalid response to fetch element via header",
+                    ?r
+                );
+            }
```

Comment:
Somehow my brain gets confused since usually there are types on the lhs, so I'm like, wait, what does `r` mean?... But this is probably personal preference, so feel free to ignore.

```suggestion
            _ => {
                error!(
                    msg = "Got an invalid response to fetch element via header",
                    ?response
                );
            }
```